### ISSUE TITLE
Shadow Dom - advertising div is not collapsing

### ISSUE THREAD

**Issue by justforfun** [ampproject/amphtml#12912]
## What's the issue?
When delivered inside a PWA through shadow dom advertising div is not collapsing

![schermata 2018-01-19 alle 11 26 55](https://user-images.githubusercontent.com/477983/35146799-6b03d19a-fd0c-11e7-8136-9379e8b16d47.png)

## How do we reproduce the issue?
It's possible to reproduce the issue here:
https://rep.repubblica.it/pwa/commento/2018/01/10/news/il_galateo_di_serracchiani_per_il_bon_ton_dei_sindaci-186240380/

## What browsers are affected?
FF always anche Chrome sometimes depending on the speed of the scrolling

## Which AMP version is affected?
Version 1516337355291
.

**Comment by aghassemi** [comment]
FF never collapsing might be a  bug, Chrome sometime not collapsing is not. We don't collapse if 1- it is in the viewport 2- may cause content shifting.

**Comment by rudygalfi** [comment]
cc @jasti

**Comment by cvializ** [comment]
@justforfun can you help us reproduce by providing credentials or instructions on how to view the full article?

**Comment by gilbertococchi** [comment]
It seems like on Firefox it happens no matter the Amp-ad element in viewport compared to the behavior on Chrome: https://rep.repubblica.it/pwa/commento/2018/01/10/news/il_galateo_di_serracchiani_per_il_bon_ton_dei_sindaci-186240380/

**Comment by jasti** [comment]
@cvializ is there an update on this please?

**Comment by cvializ** [comment]
Unfortunately the test credentials are not working now to load the article with ads.

**Comment by cvializ** [comment]
Can reproduce, no root cause found immediately, but I will investigate. /cc @danielrozenberg

**Comment by Aerlinger** [comment]
@glevitzky @zhouyx @cvializ @lannka Pinging for an update. This still seems to be an issue.

**Comment by Aerlinger** [comment]
Also @justforfun @gilbertococchi can you provide another test page? The one in this thread seems to be behind a paywall.

**Comment by justforfun** [comment]
@Aerlinger The only to reproduce the issue is to use an account authorized to access the content. I created an account for you and shared the its details with @gilbertococchi and @kul3r4 since I haven't any your direct contact.

**Comment by danielrozenberg** [comment]
I've been trying to debug this all day today, however Firefox's debugger has a few bugs that prevented me from making progress. @justforfun I was wondering if you can help by creating a minimum viable demo that causes this bug. I'll try to create one on our end this week and see if I can reproduce it by reducing the page to the key components that we think are at fault (amp-a4a.js and shadow-v0.js)

**Comment by justforfun** [comment]
@danielrozenberg : For us is not easy to create a minimum viable demo. I'm providing you the HTML code of the AMP page. In this page the ad is collapsing correctly but as soon as the page is delivered through a PWA and Polymer's <amp-viewer> the ad is not collapsing anymore. As you can easily verify the structure of the page is quite simple, you can try to reduce it by yourself in your test environment
[20180531-pwa-rep-amp.html.zip](https://github.com/ampproject/amphtml/files/2057636/20180531-pwa-rep-amp.html.zip)
Please let me know if you need any further information

**Comment by danielrozenberg** [comment]
I found one issue that hindered me, which is that the doubleclick ads don't collapse if Firefox's Tracking Protection feature is turned on (which is the default case for Private Browsing, hence why I kept seeing the ad not collapsing regardless of whether I'm seeing it inside a PWA or not.) This is an unrelated bug (filed #15843) and now that I have this out of the way I can continue to investigate.

So far I started with the simplest page imaginable (a bunch of text and at the end of it a non-existent doubleclick ad) and this ad collapses in both regular-amp-page mode and in amp-embedded-in-shadow-root mode, so I will continue to add on components and complexity from your page until I find the cause.

**Comment by danielrozenberg** [comment]
I found the cause! It crashes on this function: https://github.com/ampproject/amphtml/blob/7112c52e1bf6606d84a106498d8e199af46bb600/src/ad-helper.js#L83-L95

Because in Firefox we rename the `<body>` tag to `<amp-body>` (I'm not sure where or why, but this is the difference) and the loop guards aren't set up to catch this case. I'm working on a fix and will send it out soon. I tested the Repubblica case with a local build, and it fixes this issue, however we won't know 100% until the fix is merged and release to canary (tho I'm quite confident about it)
